Update plugin generator to generate NP plugins

lizozom opened the pull request:
## Summary

Following feedback from https://github.com/elastic/kibana/pull/52752

`node scripts/generate_plugin.js "Banana plugin"`

To generate a plugin, with the following options supported:
 * Generate FE components
 * Generate server side endpoint
 * Generate SCSS
 * Generate an optional .eslintrc.js file (for 3rd party plugins mostly)
 * Init git repo (for 3rd party plugins)

https://github.com/elastic/kibana/pull/57766 adds also
 * Support for custom folder (to generate x-pack or example plugins)
 * Translations

### What's generated

The generated plugin has server side and\or client side components.
The client side features a button that, when pressed, uses the core `http` service to make a request (or if no server side generated, just shows a notification).

The plugin uses the core `notifications` service to display a notification.

![image](https://user-images.githubusercontent.com/3016806/73443978-5fb76f80-4360-11ea-989b-690c8c2dc4e8.png)

## Dev Docs

Add more options for Kibana new platform plugin generator. 
 * Generate FE components
 * Generate server side endpoint
 * Generate SCSS
 * Generate an optional .eslintrc.js file (for 3rd party plugins mostly)
 * Init git repo (for 3rd party plugins)
### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
Pinging @elastic/kibana-operations (Team:Operations)

elasticmachine commented:
Pinging @elastic/kibana-app-arch (Team:AppArch)

lizozom commented:
@joshdover 

What other things should be initialized with a new platform plugin?

ATM it generates the following structure:

plugin 
 - common
 -- index.ts
 - public
 -- components
 --- app.ts
 -- index.ts
 -- types.ts
 -- plugin.ts
 - server
 -- index.ts
 -- plugin.ts
 kibana.json
 README.md

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/)
* Commit: 5eb53575f53707ca958ea3a16832f6e86988ce4e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/","number":"21298","commit":"5eb53575f53707ca958ea3a16832f6e86988ce4e"}]}
PIPELINE-->

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/)
* Commit: f44b2c39e7915443999f9fe8c99813927d85ad41

### History
* :broken_heart: [Build #21298](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/) failed 5eb53575f53707ca958ea3a16832f6e86988ce4e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/","number":"21298","commit":"5eb53575f53707ca958ea3a16832f6e86988ce4e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/","number":"21329","commit":"f44b2c39e7915443999f9fe8c99813927d85ad41"}]}
PIPELINE-->

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/)
* Commit: 65e621d63162d2b7747a48fdb45b0a1f40be84ae

### History
* :broken_heart: [Build #21329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/) failed f44b2c39e7915443999f9fe8c99813927d85ad41
* :broken_heart: [Build #21298](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/) failed 5eb53575f53707ca958ea3a16832f6e86988ce4e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/","number":"21298","commit":"5eb53575f53707ca958ea3a16832f6e86988ce4e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/","number":"21329","commit":"f44b2c39e7915443999f9fe8c99813927d85ad41"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/","number":"21343","commit":"65e621d63162d2b7747a48fdb45b0a1f40be84ae"}]}
PIPELINE-->

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/kibana.json:
```diff
@@ -0,0 +1,9 @@
+    "requiredPlugins": [],
+    "optionalPlugins": []
+  }
+  
```
NIT: missing linefeed on most files

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/server/plugin.ts:
```diff
@@ -0,0 +1,21 @@
+export interface <%= camelCaseName %>PluginStart {};
+
+export class <%= camelCaseName %>ServerPlugin implements Plugin<<%= camelCaseName %>PluginSetup, <%= camelCaseName %>PluginStart> {
+  constructor(initializerContext: PluginInitializerContext) {}
```
I think it would be nice to add a `Logger` property to the example plugin class and create it from the `initializerContext `

lizozom replied:
I replaced `initializerContext` with the new way of getting context: `core.getStartServices();`
To demonstrate I decided to use the notification service in the app.
I also want to use a non-core plugin. Probably top nav.

pgayvallet replied:
Example of using `getStartServices` is good. However they are not the same thing as the `initializerContext`.

I meant showing how to declare / use a logger in NP:

```typescript
private readonly logger: Logger;

constructor(private readonly initializerContext: PluginInitializerContext) {   
  this.logger = this.initializerContext.logger.get();
}

public start(core: CoreStart) {
    this.logger.debug('<watheverPluginName>: Started');
    return {};
}
```

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/server/plugin.ts:
```diff
@@ -0,0 +1,21 @@
+import { PluginInitializerContext, CoreSetup, CoreStart, Plugin } from '<%= relRoot %>/src/core/server';
+
+export interface <%= camelCaseName %>PluginSetup {};
+export interface <%= camelCaseName %>PluginStart {};
```
NIT: inconsistency: the setup/start interfaces got their own `types` file for `public` but are in the plugin file for `server`. both are fine, but public/server should have the same structure

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/public/plugin.ts:
```diff
@@ -0,0 +1,34 @@
+    return {
+      getGreeting() {
+        return 'Hello from Plugin A!';
+      },
```
This method is not defined in the associated `<>PublicPluginSetup` in `public/types.ts`

pgayvallet commented on packages/kbn-plugin-generator/index.js.d.ts:
```diff
@@ -0,0 +1,24 @@
+  /**
+   * Run plugin generator.
+   */
+  run: any;
```
NIT: should probably be at least callable: `(...args[]) => any` or something

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/public/components/app.tsx:
```diff
@@ -0,0 +1,93 @@
+              <h1>
+                <FormattedMessage
+                  id="liz.helloWorldText"
+                  defaultMessage="Hello World!"
```
NIT: bundle prefix should be something like `snake_case(pluginName)`. Also we might want to display the plugin name in the app component's title.

lizozom commented:
@pgayvallet I added a server side component to the generator.

I still need to support the `hasUi` \ `hasServer` flags - both components are generated regardless of their values.

Also, I'd like to add a non-core dependency. Not sure which one. Maybe `TopNavMenu`.

lizozom commented:
@elastic/kibana-design I would love some input on making the demo plugin's UI better (keeping it simple though!)

ryankeairns commented:
@lizozom I've added this to our design team's meeting agenda for today. We'll discuss and get back to you!

ryankeairns commented:
This is a nice onramp, easy to follow, quick to set up. In particular, I like the use of `EuiPage` components here as a starting point since that is a common pattern.

Keeping it really simple, I think we should make the following minor changes:

- Change the title and subtitle text. The "Hello.." intro is fun, but doesn't read well with unpredictable plugin names. We can also combine the subheadings in the content section... we probably shouldn't presume it is somebody "first" plugin.
- Add a descriptive paragraph with some recommendations on next steps (a link to docs would be nice, if possible).
- Limit the width of the page by adding a prop `<EuiPage restrictWidth="1000px">`. Looks a little tidier and is the common width we use on other app home pages.
- Add a horizontal rule to separate the data fetching section

The end result would like something like this:

<img width="1561" alt="Screenshot 2020-01-21 13 02 53" src="https://user-images.githubusercontent.com/446285/72834707-fc865880-3c4e-11ea-9bd0-77ebce05149e.png">

Let me know if you have questions or would like to go another route.

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/public/index.ts:
```diff
@@ -0,0 +1,10 @@
+
+// Export plugin after all other imports
+import { <%= upperCamelCaseName %>PublicPlugin } from './plugin';
+export { <%= upperCamelCaseName %>PublicPlugin as Plugin };
```
NITs: 
- the import should be at the start of the file
- do we really need to export the `Plugin` class? isn't the factory `plugin` enough? (same goes for the `server` counterpart)

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/server/plugin.ts:
```diff
@@ -0,0 +1,36 @@
+        });
+      }
+    );
+    return {};
```
Following conventions, all routes should be defined in a `routes` subfolder. (see `x-pack/plugins/security/server/routes` or `src/core/server/capabilities/routes` for exemple)

lizozom commented:
@ryankeairns addressed your feedback (don't know what link to put there ATM, but added the text paragraph anyway.)

ryankeairns commented on packages/kbn-plugin-generator/sao_template/template/public/components/app.tsx:
```diff
@@ -70,14 +74,14 @@ export const <%= upperCamelCaseName %>App = ({ basename, notifications, http, na
                     <FormattedMessage
                       id="<%= camelCase(name) %>.helloWorldText"
-                      defaultMessage="Hello {name}!"
+                      defaultMessage="{name}!"
```
The exclamation point can be removed here.

ryankeairns commented on packages/kbn-plugin-generator/sao_template/template/public/components/app.tsx:
```diff
@@ -89,19 +93,20 @@ export const <%= upperCamelCaseName %>App = ({ basename, notifications, http, na
                       <FormattedMessage
                         id="<%= camelCase(name) %>.congratulationsTitle"
-                        defaultMessage="Congratulations"
+                        defaultMessage="Congratulations, you have successfully created your first Kibana Plugin!"
```
```suggestion
                        defaultMessage="Congratulations, you have successfully created a new Kibana Plugin!"
```

ryankeairns replied:
Let's not presume it is their first plugin.

ryankeairns commented on packages/kbn-plugin-generator/sao_template/template/public/components/app.tsx:
```diff
@@ -89,19 +93,20 @@ export const <%= upperCamelCaseName %>App = ({ basename, notifications, http, na
-                        id="<%= camelCase(name) %>.congratulationsText"
-                        defaultMessage="You have successfully created your first Kibana Plugin!"
+                        id="<%= camelCase(name) %>.content"
+                        defaultMessage="Look through the generated code and checkout the plugin developmment documentation."
```
If possible, I believe we would want to point them to https://www.elastic.co/guide/en/kibana/current/plugin-development.html

@gchaps or @KOTungseth do we account for the version in some way when adding docs links to the UI? I'm thinking there might be some kind of helper in Kibana for this, but I might also be imagining things :)

ryankeairns replied:
For reference, you can see the proposed UI copy and link placement here: https://github.com/elastic/kibana/pull/55281#issuecomment-576833494

gchaps replied:
@ryankeairns, @lizozom  we do account for version.  I don't know how it is done, but we use links elsewhere in the UI.

I'm not sure how well "Fetch data" will do with localization.  Can the button text be "Get data" instead?

Also, how about shortening the title line a bit:

Congratulations, you created a Kibana plugin!

lizozom replied:
`Congratulations, you created a new Kibana plugin!` works?

lizozom replied:
Ok on text changes
I'd like to leave the doc link change to a follow up PR, if you don't mind?

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/)
* Commit: 9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366

### History
* :broken_heart: [Build #21343](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/) failed 65e621d63162d2b7747a48fdb45b0a1f40be84ae
* :broken_heart: [Build #21329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/) failed f44b2c39e7915443999f9fe8c99813927d85ad41
* :broken_heart: [Build #21298](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/) failed 5eb53575f53707ca958ea3a16832f6e86988ce4e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/","number":"21298","commit":"5eb53575f53707ca958ea3a16832f6e86988ce4e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/","number":"21329","commit":"f44b2c39e7915443999f9fe8c99813927d85ad41"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/","number":"21343","commit":"65e621d63162d2b7747a48fdb45b0a1f40be84ae"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/","number":"22601","commit":"9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366"}]}
PIPELINE-->

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/public/index.ts:
```diff
@@ -0,0 +1,9 @@
+
+import { <%= upperCamelCaseName %>PublicPlugin } from './plugin';
+export function plugin() {
```
Ultra NIT: move empty line from before to after import statement

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/server/routes/index.ts:
```diff
@@ -0,0 +1,17 @@
+    {
+      path: '/api/<%= snakeCase(name) %>/example',
+      validate: false,
+    },
```
Should we add an example usage of validation, to demonstrate how the `@kbn/config-schema` validation works?

lizozom replied:
I think we should keep the generator basic.
We will refer users to a library of sample plugins for more examples.

mistic commented on packages/kbn-plugin-generator/sao_template/template/translations/zh-CN.json:
```diff
@@ -1,84 +0,0 @@
-{
```
@lizozom are we considering to re-add a translation example to the plugin generator?

lizozom replied:
I don't know how to :)
How is this going to work in NP?

mistic replied:
@joshdover what is your opinion here?

mattkime replied:
Seems like this wouldn't work without a considerable hack.

lizozom replied:
Maybe we can add this in a follow up PR if everything else is a'ight?

lizozom commented:
@elasticmachine merge upstream

gchaps commented:
@lizozom Could you please post a screenshot with the latest text?

mattkime commented:
@lizozom What are your plans for testing?

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/)
* Commit: d11db3aa03d715c044c262f84cb043dba5b90bd3

### History
* :green_heart: [Build #22601](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/) succeeded 9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366
* :broken_heart: [Build #21343](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/) failed 65e621d63162d2b7747a48fdb45b0a1f40be84ae
* :broken_heart: [Build #21329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/) failed f44b2c39e7915443999f9fe8c99813927d85ad41
* :broken_heart: [Build #21298](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/) failed 5eb53575f53707ca958ea3a16832f6e86988ce4e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/","number":"21298","commit":"5eb53575f53707ca958ea3a16832f6e86988ce4e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/","number":"21329","commit":"f44b2c39e7915443999f9fe8c99813927d85ad41"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/","number":"21343","commit":"65e621d63162d2b7747a48fdb45b0a1f40be84ae"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/","number":"22601","commit":"9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/","number":"23039","commit":"d11db3aa03d715c044c262f84cb043dba5b90bd3"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/)
* Commit: 72792431fc4b32533cec188e8443b57abe1f2ea8

### History
* :green_heart: [Build #23039](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/) succeeded d11db3aa03d715c044c262f84cb043dba5b90bd3
* :green_heart: [Build #22601](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/) succeeded 9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366
* :broken_heart: [Build #21343](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/) failed 65e621d63162d2b7747a48fdb45b0a1f40be84ae
* :broken_heart: [Build #21329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/) failed f44b2c39e7915443999f9fe8c99813927d85ad41
* :broken_heart: [Build #21298](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/) failed 5eb53575f53707ca958ea3a16832f6e86988ce4e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21298/","number":"21298","commit":"5eb53575f53707ca958ea3a16832f6e86988ce4e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/","number":"21329","commit":"f44b2c39e7915443999f9fe8c99813927d85ad41"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/","number":"21343","commit":"65e621d63162d2b7747a48fdb45b0a1f40be84ae"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/","number":"22601","commit":"9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/","number":"23039","commit":"d11db3aa03d715c044c262f84cb043dba5b90bd3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/","number":"23298","commit":"72792431fc4b32533cec188e8443b57abe1f2ea8"}]}
PIPELINE-->

Dosant commented on packages/kbn-plugin-generator/sao_template/template/public/components/app.tsx:
```diff
@@ -0,0 +1,129 @@
+                    <p>
+                      <FormattedMessage
+                        id="<%= camelCase(name) %>.content"
+                        defaultMessage="Look through the generated code and checkout the plugin developmment documentation."
```
typo: `developmment` -> `development`

Dosant commented on packages/kbn-plugin-generator/sao_template/template/public/plugin.ts:
```diff
@@ -0,0 +1,42 @@
+    core.application.register({
+      id: '<%= camelCase(name) %>',
+      title: PLUGIN_NAME,
+      async mount(params) {
```
Here my IDE complains on params implicitly being any

Dosant commented on packages/kbn-plugin-generator/sao_template/template/public/components/app.tsx:
```diff
@@ -0,0 +1,129 @@
+                    <EuiHorizontalRule/>
+                    <p>
+                      <FormattedMessage
+                        id="banana.timestampText"
```
leftover 🍌 instead of plugin id

Dosant replied:
One more :) 
https://github.com/elastic/kibana/pull/55281/files#diff-7974db16cc96ebdadc2997afad18f7a8R60

Dosant commented on packages/kbn-plugin-generator/sao_template/template/public/components/app.tsx:
```diff
@@ -0,0 +1,129 @@
+  EuiText,
+} from '@elastic/eui';
+
+import { AppMountParameters, CoreStart } from '<%= relRoot %>/../src/core/public';
```
AppMountParameters are not used in this file. Should be removed?

lizozom commented:
@Dosant 
 - I added scss support as per your suggestion - it will work properly only after #53976 is merged.
 - I think translation files can be added later, because I'm still not sure how those will work with NP plugins.

@mattkime 
This was already covered [integration tests](packages/kbn-plugin-generator/integration_tests/generate_plugin.test.js). Those are creating a plugin, linting it and then running browser tests on the kibana with the generated plugin.
Any additional suggestions on testing?

lizozom commented:
@Dosant

> Also noticed a weird behaviour, I am not sure if this is a platform bug or something with this plugin setup:
> When you you launch a demo app - there is a flickering page "Plugin not found"

I noticed it! @joshdover is this something to expect when loading a third party NP plugin?

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/)
* Commit: 38aee60825afbef5bd97f6e3c9f27665f0d0725e

### History
* :green_heart: [Build #23298](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/) succeeded 72792431fc4b32533cec188e8443b57abe1f2ea8
* :green_heart: [Build #23039](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/) succeeded d11db3aa03d715c044c262f84cb043dba5b90bd3
* :green_heart: [Build #22601](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/) succeeded 9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366
* :broken_heart: [Build #21343](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/) failed 65e621d63162d2b7747a48fdb45b0a1f40be84ae
* :broken_heart: [Build #21329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/) failed f44b2c39e7915443999f9fe8c99813927d85ad41

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21329/","number":"21329","commit":"f44b2c39e7915443999f9fe8c99813927d85ad41"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/","number":"21343","commit":"65e621d63162d2b7747a48fdb45b0a1f40be84ae"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/","number":"22601","commit":"9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/","number":"23039","commit":"d11db3aa03d715c044c262f84cb043dba5b90bd3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/","number":"23298","commit":"72792431fc4b32533cec188e8443b57abe1f2ea8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/","number":"23330","commit":"38aee60825afbef5bd97f6e3c9f27665f0d0725e"}]}
PIPELINE-->

lukeelmers commented on packages/kbn-plugin-generator/sao_template/template/public/application.tsx:
```diff
@@ -0,0 +1,25 @@
+        basename={appBasePath}
+        notifications={notifications}
+        http={http}
+        navigation={navigation}
```
Just adding a note that we might want to revise the structure of these props later, pending the discussions in #53029

lizozom replied:
Will keep a close watch on this.

joshdover commented:
> I noticed it! @joshdover is this something to expect when loading a third party NP plugin?

Sounds like a bug. Does this only happen with 3rd party plugins? I'd be surprised to find any difference between those and 1st party ones. Either way, can you file a bug as much info as possible.

lizozom commented:
@joshdover filed: https://github.com/elastic/kibana/issues/56362
@lukeelmers addressed all of your comments 

I will address the disabled tests in this PR. 

Assuming that one is done, is this PR ok to merge in its current state?

streamich commented on packages/kbn-plugin-generator/sao_template/template/README.md:
```diff
@@ -8,32 +8,5 @@
 
 ## development
```
nit

```suggestion
## Development
```

streamich commented on packages/kbn-plugin-generator/sao_template/template/eslintrc.js:
```diff
@@ -1,24 +1,9 @@
-module.exports = {
-  root: true,
+module.exports = {	
```
When generator is run with `-i` this file `.eslintrc.js` is added to Kibana internal plugin, not sure if that is expected.

lizozom replied:
fixed

streamich commented on packages/kbn-plugin-generator/sao_template/template/public/index.ts:
```diff
@@ -0,0 +1,15 @@
+
+import { <%= upperCamelCaseName %>Plugin } from './plugin';
+
+// These exports are the public contract for your static code and types.
```
nit: maybe we should not mention "static" as when the `plugin()` is initialized by the platform it is stateful.

```suggestion
// This exports static code and TypeScript types, as well as, Kibana Platform `plugin()` initializer.
```

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23473/)
* Commit: a9d471bb5fd60589664beb51d680f1e5f24dabf3

### History
* :green_heart: [Build #23330](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/) succeeded 38aee60825afbef5bd97f6e3c9f27665f0d0725e
* :green_heart: [Build #23298](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/) succeeded 72792431fc4b32533cec188e8443b57abe1f2ea8
* :green_heart: [Build #23039](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/) succeeded d11db3aa03d715c044c262f84cb043dba5b90bd3
* :green_heart: [Build #22601](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/) succeeded 9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366
* :broken_heart: [Build #21343](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/) failed 65e621d63162d2b7747a48fdb45b0a1f40be84ae

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/21343/","number":"21343","commit":"65e621d63162d2b7747a48fdb45b0a1f40be84ae"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/","number":"22601","commit":"9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/","number":"23039","commit":"d11db3aa03d715c044c262f84cb043dba5b90bd3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/","number":"23298","commit":"72792431fc4b32533cec188e8443b57abe1f2ea8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/","number":"23330","commit":"38aee60825afbef5bd97f6e3c9f27665f0d0725e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23473/","number":"23473","commit":"a9d471bb5fd60589664beb51d680f1e5f24dabf3"}]}
PIPELINE-->

lizozom commented:
@streamich the scss is being used https://github.com/elastic/kibana/pull/55281/files#diff-880279824dc8e1e8b83ce12c3cc00bf0R2

The yarn addition is nice, but I don't want to expand the scope of this PR. 
Lets add it as we get people start using it.

lukeelmers commented on packages/kbn-plugin-generator/sao_template/sao.js:
```diff
@@ -61,13 +61,19 @@ module.exports = function({ name, targetPath, isKibanaPlugin }) {
       },
+      generateEslint: {
+        type: 'confirm',
+        message: 'Should generate custom eslint file be used?',
```
The phrasing is a little confusing here. Maybe something closer to:

```suggestion
        message: 'Would you like to use a custom eslint file?',
```

lukeelmers commented on packages/kbn-plugin-generator/sao_template/template/public/components/app.tsx:
```diff
@@ -0,0 +1,129 @@
+    http.get('/api/<%= snakeCase(name) %>/example').then(res => {
+      setTimestamp(res.time);
+      // Use the core notifications service to display a success message.
+      notifications.toasts.addSuccess(i18n.translate('banana.dataUpdated', {
```
Another leftover 🍌

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23662/)
* Commit: 0b9ca29b1bd2440a091877f4ef0e8f54e4eb2357

### History
* :broken_heart: [Build #23473](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23473/) failed a9d471bb5fd60589664beb51d680f1e5f24dabf3
* :green_heart: [Build #23330](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/) succeeded 38aee60825afbef5bd97f6e3c9f27665f0d0725e
* :green_heart: [Build #23298](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/) succeeded 72792431fc4b32533cec188e8443b57abe1f2ea8
* :green_heart: [Build #23039](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/) succeeded d11db3aa03d715c044c262f84cb043dba5b90bd3
* :green_heart: [Build #22601](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/) succeeded 9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/22601/","number":"22601","commit":"9f2bcb8bf798e4885ac7c6c6eb61ef766fd71366"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/","number":"23039","commit":"d11db3aa03d715c044c262f84cb043dba5b90bd3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/","number":"23298","commit":"72792431fc4b32533cec188e8443b57abe1f2ea8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/","number":"23330","commit":"38aee60825afbef5bd97f6e3c9f27665f0d0725e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23473/","number":"23473","commit":"a9d471bb5fd60589664beb51d680f1e5f24dabf3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23662/","number":"23662","commit":"0b9ca29b1bd2440a091877f4ef0e8f54e4eb2357"}]}
PIPELINE-->

pgayvallet commented on packages/kbn-plugin-generator/integration_tests/generate_plugin.test.js:
```diff
@@ -78,16 +79,7 @@ describe(`running the plugin-generator via 'node scripts/generate_plugin.js plug
-    it(`'yarn build' should exit 0`, async () => {
+    it.skip(`'yarn build' should exit 0`, async () => {
       await execa('yarn', ['build'], { cwd: generatedPath });
     });
```
I see the comment about the i18n test, what about the `yarn build` and `yarn lint`. If they are obsolete they should probably be removed instead of skipped?

pgayvallet commented on packages/kbn-plugin-generator/sao_template/sao.js:
```diff
@@ -47,19 +45,14 @@ module.exports = function({ name }) {
         type: 'confirm',
-        message: 'Should a server API be generated?',
+        message: 'Should translation files be generated?',
         default: true,
```
Translations are now used/handled right now, correct? should probably disable this prompt until then?

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/public/index.ts:
```diff
@@ -0,0 +1,18 @@
+ **/ 
+export function plugin() {
+  return new <%= upperCamelCaseName %>Plugin();
+}
```
This is a file description comment right? I think `/** **/ ` should be changed to `/* */` as atm it's considered as jsdoc for the `plugin` export

lizozom replied:
It's actually just a comment for the user generating the plugin.
I'm switching it to // comments to avoid confusion.

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/server/index.ts:
```diff
@@ -0,0 +1,15 @@
+ **/ 
+ export function plugin(initializerContext: PluginInitializerContext) {
+  return new <%= upperCamelCaseName %>Plugin(initializerContext);
+}
```
NIT: same than on the `public` side

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/public/types.ts:
```diff
@@ -0,0 +1,11 @@
+
+export interface AppPluginDependencies { 
+  navigation: NavigationPublicPluginStart 
+};
```
- `AppPluginDependencies` -> `AppPluginStartDependencies` (as there could also be a `AppPluginSetupDependencies` for `setup` phase)

- Should probably be `<%= upperCamelCaseName %>PluginStartDependencies`

pgayvallet commented on packages/kbn-plugin-generator/sao_template/template/public/plugin.ts:
```diff
@@ -0,0 +1,42 @@
+        // Get start services as specified in kibana.json
+        const [coreStart, depsStart] = await core.getStartServices();
+        // Render the application
+        return renderApp(coreStart, depsStart as AppPluginDependencies, params);
```
You can avoid the manual casting by specifying the plugin's dependencies:

`Plugin<<%= upperCamelCaseName %>PluginSetup, <%= upperCamelCaseName %>PluginStart>`

to 

`Plugin<<%= upperCamelCaseName %>PluginSetup, <%= upperCamelCaseName %>PluginStart>, {}, AppPluginDependencies `

joshdover replied:
Unfortunately, I think you also have to specify the generic on `CoreSetup` on line 9, because TS does not infer types for function arguments on interface implementations.

Line 9 could be:
```
public setup(core: CoreSetup<AppPluginDependencies>): <%= upperCamelCaseName %>PluginSetup {
```

joshdover commented on packages/kbn-plugin-generator/sao_template/template/package_template.json:
```diff
@@ -1,41 +0,0 @@
-{
```
Why did we decide to remove this? I think it's still useful for plugins to be able to have their own dependencies and we should set them up to do that. The yarn workspaces config we have should support installing dependencies when running `yarn kbn bootstrap` for plugins in this directory.

In addition, I think having them use our eslint config by default is a good idea since it helps prevent some common problems. However, they're not currently depending on it, this just relies on the plugin being in a specific directory. I suspect we won't need to require this once #53976 moves further along.

I also expect that they'll need to depend on `@kbn/config-schema` in order to do anything useful on the server-side.

lizozom replied:
@joshdover I removed it because I was able to create and run plugins without it. I prefer to have a minimal structure, then re-introduce things based on need.

What do you think?

joshdover replied:
I believe the primary audience for this package is 3rd parties. With the current system, there's a bit of complexity involved in setting up a package.json file to work with Kibana plugins (eg. the `link:` dependencies). It seems to me we should help them do this, at least as an option, since I imagine most plugins will want to use some dependencies of their own (though that is an assumption).

joshdover replied:
We can do this in a separate follow-up PR.

lizozom replied:
Opened an issue for addressing translations and packages.json

https://github.com/elastic/kibana/issues/56652

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23907/)
* Commit: 8ad025ec6f55e8a55cd3edb61ff6f6369b4f79a9

### History
* :green_heart: [Build #23662](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23662/) succeeded 0b9ca29b1bd2440a091877f4ef0e8f54e4eb2357
* :broken_heart: [Build #23473](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23473/) failed a9d471bb5fd60589664beb51d680f1e5f24dabf3
* :green_heart: [Build #23330](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/) succeeded 38aee60825afbef5bd97f6e3c9f27665f0d0725e
* :green_heart: [Build #23298](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/) succeeded 72792431fc4b32533cec188e8443b57abe1f2ea8
* :green_heart: [Build #23039](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/) succeeded d11db3aa03d715c044c262f84cb043dba5b90bd3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23039/","number":"23039","commit":"d11db3aa03d715c044c262f84cb043dba5b90bd3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/","number":"23298","commit":"72792431fc4b32533cec188e8443b57abe1f2ea8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/","number":"23330","commit":"38aee60825afbef5bd97f6e3c9f27665f0d0725e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23473/","number":"23473","commit":"a9d471bb5fd60589664beb51d680f1e5f24dabf3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23662/","number":"23662","commit":"0b9ca29b1bd2440a091877f4ef0e8f54e4eb2357"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23907/","number":"23907","commit":"8ad025ec6f55e8a55cd3edb61ff6f6369b4f79a9"}]}
PIPELINE-->

lizozom commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23914/)
* Commit: 8274c78783cc35bb2fca5beec9ff595ab4611d76

### History
* :green_heart: [Build #23907](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23907/) succeeded 8ad025ec6f55e8a55cd3edb61ff6f6369b4f79a9
* :green_heart: [Build #23662](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23662/) succeeded 0b9ca29b1bd2440a091877f4ef0e8f54e4eb2357
* :broken_heart: [Build #23473](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23473/) failed a9d471bb5fd60589664beb51d680f1e5f24dabf3
* :green_heart: [Build #23330](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/) succeeded 38aee60825afbef5bd97f6e3c9f27665f0d0725e
* :green_heart: [Build #23298](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/) succeeded 72792431fc4b32533cec188e8443b57abe1f2ea8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23298/","number":"23298","commit":"72792431fc4b32533cec188e8443b57abe1f2ea8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/","number":"23330","commit":"38aee60825afbef5bd97f6e3c9f27665f0d0725e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23473/","number":"23473","commit":"a9d471bb5fd60589664beb51d680f1e5f24dabf3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23662/","number":"23662","commit":"0b9ca29b1bd2440a091877f4ef0e8f54e4eb2357"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23907/","number":"23907","commit":"8ad025ec6f55e8a55cd3edb61ff6f6369b4f79a9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23914/","number":"23914","commit":"8274c78783cc35bb2fca5beec9ff595ab4611d76"}]}
PIPELINE-->

streamich commented:
@lizozom the suggestion regarding SASS was to create an `index.scss` file next to every React component and import it from that React component.

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/27144/)
* Commit: 8274c78783cc35bb2fca5beec9ff595ab4611d76

### History
* :green_heart: [Build #23914](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23914/) succeeded 8274c78783cc35bb2fca5beec9ff595ab4611d76
* :green_heart: [Build #23907](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23907/) succeeded 8ad025ec6f55e8a55cd3edb61ff6f6369b4f79a9
* :green_heart: [Build #23662](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23662/) succeeded 0b9ca29b1bd2440a091877f4ef0e8f54e4eb2357
* :broken_heart: [Build #23473](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23473/) failed a9d471bb5fd60589664beb51d680f1e5f24dabf3
* :green_heart: [Build #23330](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/) succeeded 38aee60825afbef5bd97f6e3c9f27665f0d0725e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23330/","number":"23330","commit":"38aee60825afbef5bd97f6e3c9f27665f0d0725e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23473/","number":"23473","commit":"a9d471bb5fd60589664beb51d680f1e5f24dabf3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23662/","number":"23662","commit":"0b9ca29b1bd2440a091877f4ef0e8f54e4eb2357"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23907/","number":"23907","commit":"8ad025ec6f55e8a55cd3edb61ff6f6369b4f79a9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/23914/","number":"23914","commit":"8274c78783cc35bb2fca5beec9ff595ab4611d76"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/27144/","number":"27144","commit":"8274c78783cc35bb2fca5beec9ff595ab4611d76"}]}
PIPELINE-->
